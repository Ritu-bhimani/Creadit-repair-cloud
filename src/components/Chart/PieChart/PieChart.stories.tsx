import { ComponentMeta, ComponentStory } from '@storybook/react';
import { ChartStyles } from '../styles';
import { PieChartComponent } from './PieChart';

export default {
    title: 'Pie chart',
    component: PieChartComponent
} as ComponentMeta<typeof PieChartComponent>;

const Template: ComponentStory<typeof PieChartComponent> = (args) => (
    <PieChartComponent {...args} />
);
type CustomToolTipProps = {
    active?: boolean;
    payload?: any;
    label?: string;
    cursor?: any;
    style?: any;
}
const CustomTooltip = (props: CustomToolTipProps) => {
    if (props.active && props.payload && props.payload.length) {
        return (
            <div css={ChartStyles.customTooltip} >
                <p className="label">{`${props.payload[0].value}-${props.payload[0].name}`}</p>
            </div>
        );
    }

    return null;
};

export const Primary = Template.bind({});

Primary.args = {
    data: [
        { name: 'Android', value: 10 },
        { name: 'Apple iOS', value: 20 },
        { name: 'Blackberry', value: 25 },
        { name: 'Windows Phone', value: 33 },
        { name: 'Others', value: 10 }
    ],

    colors: ['#0088FE', '#00C49F', '#FFBB28', '#FF8042', '#a5db89'],
    height: 200,
    width: 500,
    customTooltip: <CustomTooltip />,
    xAxisDataKey: "name",
    pieDataKey: "value",
};
