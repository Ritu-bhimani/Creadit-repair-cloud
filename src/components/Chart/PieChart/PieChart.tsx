import { Container } from '@mui/material';
import { FC, useCallback, useState } from 'react';
import { Cell, Legend, Pie, PieChart, Sector, Tooltip } from 'recharts';
import {
  NameType,
  ValueType
} from 'recharts/types/component/DefaultTooltipContent';
import { ContentType } from 'recharts/types/component/Tooltip';
import { ChartStyles } from '../styles';
type PieChartProps = {
  data?: any;
  height?: number;
  width?: number;
  colors?: any;
  customTooltip?: ContentType<ValueType, NameType>;
  xAxisDataKey?: string;
  pieDataKey?: any;
};

type PieChartDetails = {
  name: string;
  value?: number;
};

export const PieChartComponent: FC<PieChartProps> = (props: PieChartProps) => {
  const {
    data,
    height,
    width,
    colors,
    customTooltip,
    xAxisDataKey,
    pieDataKey
  } = props;
  const [activeIndex, setActiveIndex] = useState();
  const handleActiveShape = useCallback(
    (data: any, index: any) => {
      if (activeIndex == index) {
        setActiveIndex(undefined);
      } else {
        setActiveIndex(index);
      }
    },
    [activeIndex]
  );
  const RADIAN = Math.PI / 180;
  const renderActiveShape = (props: {
    cx: any;
    cy: any;
    innerRadius: any;
    outerRadius: any;
    startAngle: any;
    endAngle: any;
    midAngle: any;
    fill: any;
  }) => {
    const RADIAN = Math.PI / 180;
    const {
      cx,
      cy,
      innerRadius,
      outerRadius,
      startAngle,
      endAngle,
      midAngle,
      fill
    } = props;
    const sin = Math.sin(-RADIAN * midAngle);
    const cos = Math.cos(-RADIAN * midAngle);
    const sx = cx + (outerRadius - 40) * cos;
    const sy = cy + (outerRadius - 40) * sin;
    return (
      <Sector
        cx={sx}
        cy={sy}
        innerRadius={innerRadius}
        outerRadius={outerRadius}
        startAngle={startAngle}
        endAngle={endAngle}
        fill={fill}
      />
    );
  };
  const pieDetails = data.map((details: PieChartDetails)=>{ return {...details, name: `${details.name} (${details.value})`} })

  return (
    <Container css={ChartStyles.ChartCss}>
      <PieChart
        data={pieDetails}
        height={height ? height : 200}
        width={width ? width : 500}
      >
        <Legend
          layout="vertical"
          align="right"
          verticalAlign="middle"
          wrapperStyle={{ overflowY: 'auto', height: 'auto', top:'30px'}}
        />
        <Pie
          activeIndex={activeIndex}
          data={pieDetails}
          cx="50%"
          cy="50%"
          fontSize={'11px'}
          labelLine={false}
          outerRadius={80}
          fill="#8884d8"
          dataKey={pieDataKey}
          innerRadius={20}
          activeShape={renderActiveShape}
          onClick={handleActiveShape}
          stroke=""
        >
          {pieDetails.map((entry: any, index: number) => (
            <Cell key={`cell-${index}`} fill={colors[index % colors.length]} />
          ))}
        </Pie>
        <Tooltip
          content={customTooltip}
          cursor={{ fill: 'transparent' }}
          itemStyle={{ borderRadius: '0' }}
        />
      </PieChart>
    </Container>
  );
};

export default PieChartComponent;
