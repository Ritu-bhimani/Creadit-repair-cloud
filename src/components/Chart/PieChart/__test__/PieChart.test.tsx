import { render } from '@testing-library/react';
import { PieChartComponent } from '../PieChart';
it('should render Pie chart and match the snapshot', () => {
  const component = render(<PieChartComponent data={[]} />);
  expect(component).toMatchSnapshot();
});
