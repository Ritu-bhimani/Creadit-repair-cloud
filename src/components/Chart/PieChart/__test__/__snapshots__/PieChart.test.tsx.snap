// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Pie chart and match the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      >
        <div
          class="recharts-wrapper"
          role="region"
          style="position: relative; cursor: default; width: 500px; height: 200px;"
        >
          <svg
            class="recharts-surface"
            cx="50%"
            cy="50%"
            height="200"
            viewBox="0 0 500 200"
            width="500"
          >
            <title />
            <desc />
            <defs>
              <clippath
                id="recharts1-clip"
              >
                <rect
                  height="190"
                  width="490"
                  x="5"
                  y="5"
                />
              </clippath>
            </defs>
          </svg>
          <div
            class="recharts-legend-wrapper"
            style="position: absolute; width: auto; height: auto; right: 5px; overflow-y: auto; top: 30px;"
          />
          <div
            class="recharts-tooltip-wrapper"
            role="dialog"
            style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
            tabindex="-1"
          >
            <div
              class="recharts-default-tooltip"
              style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
            >
              <p
                class="recharts-tooltip-label"
                style="margin: 0px;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
    >
      <div
        class="recharts-wrapper"
        role="region"
        style="position: relative; cursor: default; width: 500px; height: 200px;"
      >
        <svg
          class="recharts-surface"
          cx="50%"
          cy="50%"
          height="200"
          viewBox="0 0 500 200"
          width="500"
        >
          <title />
          <desc />
          <defs>
            <clippath
              id="recharts1-clip"
            >
              <rect
                height="190"
                width="490"
                x="5"
                y="5"
              />
            </clippath>
          </defs>
        </svg>
        <div
          class="recharts-legend-wrapper"
          style="position: absolute; width: auto; height: auto; right: 5px; overflow-y: auto; top: 30px;"
        />
        <div
          class="recharts-tooltip-wrapper"
          role="dialog"
          style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
          tabindex="-1"
        >
          <div
            class="recharts-default-tooltip"
            style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
          >
            <p
              class="recharts-tooltip-label"
              style="margin: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
